in-progress/11103-forwarder-contract/design.md

```diff
@@ -0,0 +1,203 @@
+| -------------------- | ---------------------------------------------------------------------------------- |
+| Issue                | [Forwarder Contract](https://github.com/AztecProtocol/aztec-packages/issues/11103) |
+| Owners               | @just-mitch                                                                        |
+| Approvers            | @LHerskind @PhilWindle @spalladino @spypsy                                         |
+| Target Approval Date | 2025-01-15                                                                         |
+
+## Executive Summary
+
+Add a forwarder contract that allows the sequencer client to take multiple actions in the same L1 transaction.
+
+Adjust the sequencer client to batch its actions into a single L1 transaction.
+
+## Introduction
+
+Within the same L1 transaction, one cannot make blob transactions and regular transactions from the same address.
+
+However, aztec node operators must be able to do things like:
+
+- propose and l2 block
+- vote in the governance proposer contract
+- claim an epoch proof quote
+  all in the same L1 block.
+
+### Goals
+
+- Allow the sequencer client to take multiple actions in the same L1 transaction
+- No changes to governance/staking
+- Under 50 gas overhead per L2 transaction when operating at 10TPS
+
+### Non-goals
+
+- Support multiple actions for the prover node
+
+## Interface
+
+Node operators will need to deploy a forwarder contract.
+
+When an attester deposits into the staking contract, the forwarder contract of the node operator will be specified as the proposer.
+
+The Aztec Labs sequencer client implementation will need to be updated to use the forwarder contract; this involves refactoring `yarn-project/sequencer-client/src/publisher/l1-publisher.ts`.
+
+## Implementation
+
+### Forwarder Contract
+
+It is straightforward.
+
+```solidity
+import {Address} from "@oz/utils/Address.sol";
+import {Ownable} from "@oz/access/Ownable.sol";
+
+contract Forwarder is Ownable {
+  using Address for address;
+
+  constructor(address __owner) Ownable(__owner) {}
+
+  function forward(address[] calldata _to, bytes[] calldata _data) external onlyOwner {
+    require(_to.length == _data.length);
+    for (uint256 i = 0; i < _to.length; i++) {
+      _to[i].functionCall(_data[i]);
+    }
+  }
+}
+```
+
+Note: this requires all the actions to succeed, so the sender must be sure that, e.g. a failed governance vote will not prevent the L2 block from being proposed.
+
+Note: this implementation is not technically part of the protocol, and as such will live in `l1-contracts/src/periphery`.
+
+### Refactoring L1 Publisher
+
+L1 publisher will be broken into two classes:
+
+- within `@aztec/sequencer-client`, there will be a `L1TxManager`
+- within `@aztec/prover-node`, there will be a `L1TxPublisher`
+
+Under the hood, both of these will use the `L1TxUtils` to create and send L1 transactions.
```

Comment:
Note that the `L1Publisher` is being used also for querying (we were lazy, yeah), which should not be among its responsibilities. I added a wrapper around the Rollup contract for querying some info at `yarn-project/ethereum/src/contracts/rollup.ts`, maybe we can move all getters there?